### PR TITLE
Interactive downsampling large vector fields (example gallery)

### PR DESCRIPTION
**Pull request by iuryt**
I made a promise almost a year ago to create this example and here I am.
The idea came from those discussions from discourse.
https://discourse.holoviz.org/t/highres-quiver-plot-returning-memory-error/6261/26
https://discourse.holoviz.org/t/dealing-high-resolution-vector-plots-xarray-holoviz/6310

Direct plotting large vector fields can quickly consume excessive memory, leading to `MemoryError`. To address this issue, we introduce a dynamic downsampling technique that enables interactive exploration of vector fields without sacrificing performance. 

fd3a258f5d11b3ca780f190ed4a1af01d2f67f19

### CODE REVIEW

**Comment by maximlt** [comment]
Hi @iuryt, it looks like pre-commit has caught that the notebook hasn't been cleaned up before being pushed to Github. You need to install pre-commit as described in the [developer guide](https://hvplot.holoviz.org/developer_guide/index.html)  (and run it manually with `pre-commit run --all` in this case, the next times you commit locally won't to do that).

**Comment by iuryt** [comment]
Sorry. Gonna fix that.

**Comment by iuryt** [comment]
@maximlt 
Let me know if this is fine now. Thanks!

File: doc/reference/xarray/vectorfield.ipynb [hunk 1]
```diff
@@ -113,6 +114,104 @@
     "ds.hvplot.vectorfield(x='x', y='y', angle='angle', mag='mag',\n",
     "                      hover=False, geo=True, coastline=True)"
    ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "# Large Data"
```
Comment (maximlt) [inline doc/reference/xarray/vectorfield.ipynb:122]:
```suggestion
    "## Large Data"
```

File: doc/reference/xarray/vectorfield.ipynb [hunk 2]
```diff
@@ -113,6 +114,104 @@
...
+    "        sub = sub.coarsen(x=ix, y=iy, side=\"center\", boundary=\"trim\").mean()\n",
+    "\n",
+    "    # Create the vector field plot\n",
+    "    quiver = sub.hvplot.vectorfield(\n",
+    "        x=\"x\",\n",
+    "        y=\"y\",\n",
+    "        mag=\"mag\",\n",
+    "        angle=\"angle\",\n",
+    "        hover=False,\n",
+    "    ).opts(magnitude=\"mag\")\n",
```
Comment (maximlt) [inline doc/reference/xarray/vectorfield.ipynb:204]: Is the `.opts` call required here? Looks redundant with `mag='mag'` in the hvPlot call.

File: doc/reference/xarray/vectorfield.ipynb [hunk 3]
```diff
@@ -113,6 +114,104 @@
...
+    "\n",
+    "    return quiver\n",
+    "\n",
+    "\n",
+    "# Create interactive plot components\n",
+    "range_xy = hv.streams.RangeXY()               # Stream to capture range changes\n",
+    "filtered = hv.DynamicMap(create_quiver, streams=[range_xy])  # Dynamic plot\n",
+    "range_xy.source = filtered                   # Link stream to the plot\n",
```
Comment (maximlt) [inline doc/reference/xarray/vectorfield.ipynb:212]: Is this line required? I'd assume the stream is linked when the DynamicMap is instantiated?

File: doc/reference/xarray/vectorfield.ipynb [hunk 4]
```diff
@@ -113,6 +114,104 @@
...
+    "        angle=\"angle\",\n",
+    "        hover=False,\n",
+    "    ).opts(magnitude=\"mag\")\n",
+    "\n",
+    "    return quiver\n",
+    "\n",
+    "\n",
+    "# Create interactive plot components\n",
+    "range_xy = hv.streams.RangeXY()               # Stream to capture range changes\n",
+    "filtered = hv.DynamicMap(create_quiver, streams=[range_xy])  # Dynamic plot\n",
```
Comment (maximlt) [inline doc/reference/xarray/vectorfield.ipynb:211]: Name it perhaps `downsampled_quiver`?

File: doc/reference/xarray/vectorfield.ipynb [hunk 5]
```diff
@@ -113,6 +114,104 @@
...
+    "If we just try to call `ds.hvplot.vectorfield` this is probably returning `MemoryError`. The alternative is to dynamically downsample the data based on the visible range. This helps manage memory consumption when dealing with large datasets, especially when plotting vector fields."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "def create_quiver(x_range=None, y_range=None, nmax=10):\n",
```
Comment (maximlt) [inline doc/reference/xarray/vectorfield.ipynb:162]: Name it perhaps `downsample_quiver`?

File: doc/reference/xarray/vectorfield.ipynb [hunk 6]
```diff
@@ -113,6 +114,104 @@
...
+    "                 'angle': xr.DataArray(angle, dims=('y', 'x'), coords={'y': ys, 'x': xs})}, \n",
+    "                attrs={'crs': crs})\n",
+    "ds"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "If we just try to call `ds.hvplot.vectorfield` this is probably returning `MemoryError`. The alternative is to dynamically downsample the data based on the visible range. This helps manage memory consumption when dealing with large datasets, especially when plotting vector fields."
```
Comment (maximlt) [inline doc/reference/xarray/vectorfield.ipynb:153]: I would be nice to link to a HoloViews user guide and explain that we're going to use some HoloViews concepts to work around the memory error, while stile using hvPlot to create the quiver.

**Review by maximlt — COMMENTED** [review]
Left a few suggestions, very nice addition!

File: doc/reference/xarray/vectorfield.ipynb (hunk 2 above)
Comment (maximlt) [reply doc/reference/xarray/vectorfield.ipynb:204]: It is :)

File: doc/reference/xarray/vectorfield.ipynb (hunk 3 above)
Comment (maximlt) [reply doc/reference/xarray/vectorfield.ipynb:212]: It's not required.

File: doc/reference/xarray/vectorfield.ipynb (hunk 5 above)
Comment (maximlt) [reply doc/reference/xarray/vectorfield.ipynb:162]: Renamed

**Comment by iuryt** [comment]
Thanks @maximlt !
Sorry for not wrapping this up before.

**Pull request merged**